platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/style/sources/CustomGeometrySource.java

```diff
@@ -177,41 +219,91 @@ public boolean equals(Object object) {
+        TileID other = (TileID) object;
         return this.z == other.z && this.x == other.x && this.y == other.y;
       }
       return false;
     }
   }
 
-  private static class GeometryTileRequest implements Runnable {
-    private TileID id;
-    private GeometryTileProvider provider;
-    private WeakReference<CustomGeometrySource> sourceRef;
-    private AtomicBoolean cancelled;
+  static class GeometryTileRequest implements Runnable {
+    private final TileID id;
+    private final GeometryTileProvider provider;
+    private final Map<TileID, LinkedBlockingQueue<GeometryTileRequest>> overlapping;
+    private final Map<TileID, AtomicBoolean> inProgress;
+    private final WeakReference<CustomGeometrySource> sourceRef;
+    private final AtomicBoolean cancelled;
 
-    public GeometryTileRequest(TileID _id, GeometryTileProvider p,
-                               CustomGeometrySource _source, AtomicBoolean _cancelled) {
+    GeometryTileRequest(TileID _id, GeometryTileProvider p,
+                        Map<TileID, LinkedBlockingQueue<GeometryTileRequest>> overlapping,
+                        Map<TileID, AtomicBoolean> m,
+                        CustomGeometrySource _source, AtomicBoolean _cancelled) {
       id = _id;
       provider = p;
+      this.overlapping = overlapping;
+      inProgress = m;
       sourceRef = new WeakReference<>(_source);
       cancelled = _cancelled;
     }
 
     public void run() {
-      if (isCancelled()) {
-        return;
+      synchronized (overlapping) {
+        synchronized (inProgress) {
+          if (overlapping.get(id) == null) {
+            overlapping.put(id, new LinkedBlockingQueue<>());
+          }
+
+          if (inProgress.put(id, cancelled) != null) {
+            // request targeting this tile id is already being processed,
+            // scenario that should occur only if the tile is being requested when
+            // another request is switching threads to execute or is in executor's queue
+            overlapping.get(id).offer(this);
+            return;
+          }
+        }
       }
 
-      FeatureCollection data = provider.getFeaturesForBounds(LatLngBounds.from(id.z, id.x, id.y), id.z);
-      CustomGeometrySource source = sourceRef.get();
-      if (!isCancelled() && source != null && data != null)  {
-        source.setTileData(id, data);
+      if (!isCancelled()) {
+        FeatureCollection data = provider.getFeaturesForBounds(LatLngBounds.from(id.z, id.x, id.y), id.z);
+        CustomGeometrySource source = sourceRef.get();
+        if (!isCancelled() && source != null && data != null) {
+          source.setTileData(id, data);
+        }
+      }
+
+      synchronized (overlapping) {
+        synchronized (inProgress) {
+          inProgress.remove(id);
+          // executing the next request targeting the same tile or cleaning up if none waiting
+          GeometryTileRequest queuedRequest = overlapping.get(id).poll();
```

Comment:
Doesn't look like you are using the blocking portion of the queue.  Since everything is being done in a synchronized block, probably more efficient to use a non thread safe queue.

Reply:
Great catch, thanks @mkrussel77.